feat: Add support for excluded_attributes in JSON source hierarchy

swngarg-lt opened the pull request.

swngarg-lt committed:
added support for excluded_attributes in JSON source hierarchy

swngarg-lt committed:
changed variable name to camel case

swngarg-lt committed:
added comment

swngarg-lt commented:
![Screenshot 2024-11-08 at 7 59 53 PM](https://github.com/user-attachments/assets/25164362-8881-4463-844e-d0a47e9b5c72)
@mykola-mokhnach Seems like unit test failing due to invalid xcode path. Can you help?

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.h:
```diff
@@ -29,7 +29,7 @@ NS_ASSUME_NONNULL_BEGIN
  Return application elements tree in form of nested dictionaries
  */
-- (NSDictionary *)fb_tree;
+- (NSDictionary *)fb_tree:(NSArray *) excludedAttributes;
```
Please update the docstring and add array item types as well as possible nullability

mykola-mokhnach replied:
what are possible values for this array? I assume a set would be a better choice

swngarg-lt replied:
addressed the comment, also used Set instead of Array

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -167,7 +167,7 @@ - (NSDictionary *)fb_accessibilityTree
 }
 
-+ (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot recursive:(BOOL)recursive
++ (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot recursive:(BOOL)recursive excludedAttributes:(NSArray *) excludedAttributes
```
Please format long strings, so different args are aligned to new lines

swngarg-lt replied:
Done

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -177,11 +177,36 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot recursi
-  info[@"isAccessible"] = [@([wrappedSnapshot isWDAccessible]) stringValue];
-  info[@"isFocused"] = [@([wrappedSnapshot isWDFocused]) stringValue];
+  
+  NSDictionary *attributeBlocks = @{
```
Please add explicit dictionary types

swngarg-lt replied:
add key, value types.

mykola-mokhnach commented on WebDriverAgentTests/IntegrationTests/XCUIApplicationHelperTests.m:
```diff
@@ -40,7 +40,7 @@ - (void)testQueringSpringboard
 - (void)testApplicationTree
 {
-  XCTAssertNotNil(self.testedApplication.fb_tree);
+  XCTAssertNotNil([self.testedApplication fb_tree:@[]]);
```
Please add a proper test that verifies actual attributes exclusion

swngarg-lt replied:
added an attribute for exclusion

swngarg-lt committed:
addressed PR comments

swngarg-lt committed:
fixed test input as well

swngarg-lt committed:
change array to nil

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.h:
```diff
@@ -27,9 +27,10 @@ NS_ASSUME_NONNULL_BEGIN
 
 /**
- Return application elements tree in form of nested dictionaries
+  @Param excludedAttributes which needs to be excluded from the response possible values contain frame, enabled, visible, accessible, focused
```
please follow the same docstring format as in other methods

mykola-mokhnach replied:
please describe how the method behaves if nil is passed as the value

swngarg-lt replied:
there's already a `nil` check when this variable is used further. 

`if (excludedAttributes == nil || ![excludedAttributes containsObject:key]) {`

in that case, all the attributes will be included in the response.

swngarg-lt replied:
changed docstring format

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -177,11 +179,37 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot recursi
+  };
+
+  for (NSString *key in attributeBlocks) {
+      if (excludedAttributes == nil || ![excludedAttributes containsObject:key]) { // check for attribute exclusion
```
the comment is not needed

swngarg-lt replied:
Removed

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -191,7 +219,7 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot recursi
     info[@"children"] = [[NSMutableArray alloc] init];
     for (id<FBXCElementSnapshot> childSnapshot in childElements) {
-      [info[@"children"] addObject:[self dictionaryForElement:childSnapshot recursive:YES]];
+      [info[@"children"] addObject:[self dictionaryForElement:childSnapshot recursive:YES excludedAttributes:excludedAttributes]];
```
please format this long method call

swngarg-lt replied:
reformatted the method call

mykola-mokhnach commented on WebDriverAgentTests/IntegrationTests/XCUIApplicationHelperTests.m:
```diff
@@ -40,7 +40,7 @@ - (void)testQueringSpringboard
 - (void)testApplicationTree
 {
-  XCTAssertNotNil(self.testedApplication.fb_tree);
+  XCTAssertNotNil([self.testedApplication fb_tree:[NSSet setWithArray:@[@"visible"]]]);
```
unfortunately this test does not fully verify the new functionality. There must be an assertion that the attribute is not present in the resulting tree

swngarg-lt replied:
added a separate assertion for this case.

swngarg-lt committed:
removed comment and formatted the long method call

swngarg-lt committed:
added assertion to verify exclusion of attribute

swngarg-lt committed:
changed docstring format

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.h:
```diff
@@ -27,9 +27,10 @@ NS_ASSUME_NONNULL_BEGIN
 
 /**
- Return application elements tree in form of nested dictionaries
+  @param excludedAttributes Set of possible attributes to be excluded i.e frame, enabled, visible, accessible, focused
```
Please add: `If set to nil or an empty array then no attributes will be excluded from the resulting JSON`

swngarg-lt replied:
done

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -379,7 +409,7 @@ - (BOOL)fb_dismissKeyboardWithKeyNames:(nullable NSArray<NSString *> *)keyNames
     
     id<FBXCElementSnapshot> elementSnapshot = [extractedElement fb_takeSnapshot];
-    NSDictionary *elementAttributes = elementSnapshot ? [self.class dictionaryForElement:elementSnapshot recursive:NO] : @{};
+    NSDictionary *elementAttributes = elementSnapshot ? [self.class dictionaryForElement:elementSnapshot recursive:NO excludedAttributes:nil] : @{};
```
Please reformat this ternary operator same way it was done below for excludedAttributes

swngarg-lt replied:
done

mykola-mokhnach commented on WebDriverAgentTests/IntegrationTests/XCUIApplicationHelperTests.m:
```diff
@@ -40,7 +40,12 @@ - (void)testQueringSpringboard
-  XCTAssertNotNil(self.testedApplication.fb_tree);
+  XCTAssertNotNil([self.testedApplication fb_tree:nil]);
+  
+  NSDictionary *applicationTree = [self.testedApplication fb_tree:[NSSet setWithArray:@[@"visible"]]];
```
Please extract this check into a separate test: testApplicationTreeAttributesFiltering

swngarg-lt replied:
done

swngarg-lt committed:
PR comments resolved for reformatting, separate test case

swngarg-lt commented:
@mykola-mokhnach addressed the comments in latest fix. Still workflow is failing due to xcode developer path issue. Can you help? 

Error in failed workflows : `xcode-select: error: invalid developer directory '/Applications/Xcode_14.3.1.app/Contents/Developer' `

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -177,11 +179,37 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot recursi
+          }
+      }
+  }
+
```
Please remove extra empty lines

swngarg-lt replied:
done

mykola-mokhnach commented on WebDriverAgentTests/IntegrationTests/XCUIApplicationHelperTests.m:
```diff
@@ -40,10 +40,22 @@ - (void)testQueringSpringboard
-  XCTAssertNotNil(self.testedApplication.fb_tree);
+  XCTAssertNotNil([self.testedApplication fb_tree:nil]);
+  
+  NSDictionary *applicationTree = [self.testedApplication fb_tree:[NSSet setWithArray:@[@"visible"]]];
```
Please remove duplicated lines from this test

swngarg-lt replied:
removed

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.h:
```diff
@@ -27,9 +27,10 @@ NS_ASSUME_NONNULL_BEGIN
+  @param excludedAttributes Set of possible attributes to be excluded i.e frame, enabled, visible, accessible, focused. If set to nil or an empty array then no attributes will be excluded from the resulting JSON
+  @return application elements tree in form of nested dictionaries
  */
-- (NSDictionary *)fb_tree;
```
Please keep the original method signature to maintain the backward compatibility. Internally it should simply call this one with `excludedAttributes` set to nil

swngarg-lt replied:
added backward compatibility

swngarg-lt committed:
removed duplicate lines, empty spaces. maintain the backward compatibility

swngarg-lt committed:
added backward compatibility in test cases as well

swngarg-lt committed:
updated docstring to match the previous one.

swngarg-lt committed:
updated docstring

mykola-mokhnach commented:
> @mykola-mokhnach addressed the comments in latest fix. Still workflow is failing due to xcode developer path issue. Can you help?
> 
> Error in failed workflows : `xcode-select: error: invalid developer directory '/Applications/Xcode_14.3.1.app/Contents/Developer' `

This looks like Azure CI issue. Please ignore it for now

swngarg-lt commented:
> > @mykola-mokhnach addressed the comments in latest fix. Still workflow is failing due to xcode developer path issue. Can you help?
> > Error in failed workflows : `xcode-select: error: invalid developer directory '/Applications/Xcode_14.3.1.app/Contents/Developer' `
> 
> This looks like Azure CI issue. Please ignore it for now

Ok. Functional Tests are failing, one of the reason, it couldn't find xcode version on CI. another is a 5 min timeout issue and it seems Simulator couldn't start within 5 min.

mykola-mokhnach commented:
> > > @mykola-mokhnach addressed the comments in latest fix. Still workflow is failing due to xcode developer path issue. Can you help?
> > > Error in failed workflows : `xcode-select: error: invalid developer directory '/Applications/Xcode_14.3.1.app/Contents/Developer' `
> > 
> > 
> > This looks like Azure CI issue. Please ignore it for now
> 
> Ok. Functional Tests are failing, one of the reason, it couldn't find xcode version on CI. another is a 5 min timeout issue and it seems Simulator couldn't start within 5 min.

These are also issues related to the CI slowness. Unfortunately we cannot address them now because of a known WDIO issue. See https://github.com/webdriverio/webdriverio/issues/13778 for more details.

christian-bromann commented:
> Unfortunately we cannot address them now because of a known WDIO issu

This issue should be resolved now.
